src/d/parser/expression.d

```diff
@@ -1047,3 +1050,78 @@ StringLiteral parseStringLiteral(ref TokenRange trange) {
 	
 	return new StringLiteral(location, name.toString(trange.context));
 }
+
+/**
+ * Parse floating point literals
+ */
+FloatLiteral parseFloatLiteral(ref TokenRange trange) {
+	const location = trange.front.location;
+	auto litString = trange.front.toString(trange.context);
+
+	trange.match(TokenType.FloatLiteral);
+	import d.common.builtintype : BuiltinType;
+
+	//Look for a suffix
+	if (litString.length > 1)
+	{
+		switch (litString[$-1])
+		{
+			// https://dlang.org/spec/lex.html#FloatSuffix
+			import std.conv : to;
+			case 'f':
+			case 'F':
+				const float f = litString[0..$-1].to!float;
+				return new FloatLiteral(location, f, BuiltinType.Float);
+			case 'L':
+				import source.exception;
+				throw new CompileException(location, "SDC does not support real literals yet");
+			default:
+				//Lexed correctly but no suffix, it's a double
+				const double d = litString[0..$].to!float;
+				return new FloatLiteral(location, d, BuiltinType.Double);
+		}
+	}
+	assert(0);
+}
+
+@("Test FloatLiteral parsing")
+unittest
+{
+	import util.testinfra;
+	import source.context;
+	auto context = new Context();
+	import source.parserutil;
+	auto tokensFromString(string s, bool skipBegin = true) {
+		import source.name;
+		auto base = context.registerMixin(Location.init, s ~ '\0');
+		auto x = lex(base, context);
+		if (skipBegin)
+		{
+			x.match(TokenType.Begin);
+		}
+		return x;
+	}
+	void floatRoundTrip(T)(const string floatString, const T floatValue)
+		if (__traits(isFloating, T))
+	{
+		import d.common.builtintype : BuiltinType;
+		const BuiltinType expectedType = is(T == float) ? BuiltinType.Float : BuiltinType.Double;
+		auto tr = tokensFromString(floatString);
+		const result = parseExpression(tr);
+		if (const FloatLiteral fl = cast(FloatLiteral) result)
+		{
+			import std.exception;
+			import std.math : isClose;
+			enforce(fl.type.builtin == expectedType);
+			//Not a huge fan of this choice of maxRelDiff
+			enforce(fl.value.isClose(floatValue, float.epsilon));
+		} else {
+			import std.format;
+			throw new Exception(format("Got a %s from `%s`", typeid(result).toString(), floatString));
+		}
+	}
+	// A few values, note that "-3.14f" is a UnaExp not a floating point literal
+	floatRoundTrip("3.14f", 3.14f);
+	floatRoundTrip("420.0", 420.0);
+	floatRoundTrip("0.00001", 0.00001);
```

Comment:
You want to test the error cases. What happen with the L suffix, or when the token is not a float literal.

Reply:
You also want to check all kind of floats formats, including hex and whatnot.

Reply:
I wanted to get the scaffolding done right first, but yes.